### ISSUE TITLE
Convolution2D for same border_mode on GPU reduces data size 

### ISSUE THREAD

**Issue by MinhazPalasara** [fchollet/keras#1092]
# Run this code. I have modified the mnist_cnn.py file to have the "same" border_mode. At the end, #it displays the size of first layer output. It should be same as the input i.e (28, 28) but it is ( 27, 27)

from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation, Flatten
from keras.layers.convolutional import Convolution2D, MaxPooling2D
from keras.utils import np_utils
import theano

batch_size = 10
nb_classes = 10
nb_epoch = 0

img_rows, img_cols = 28, 28
nb_filters = 32
nb_pool = 2
nb_conv = 2

(X_train, y_train), (X_test, y_test) = mnist.load_data()

X_train = X_train.reshape(X_train.shape[0], 1, img_rows, img_cols)
X_test = X_test.reshape(X_test.shape[0], 1, img_rows, img_cols)
X_train = X_train.astype("float32")
X_test = X_test.astype("float32")
X_train /= 255
X_test /= 255

print('X_train shape:', X_train.shape)
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

Y_train = np_utils.to_categorical(y_train, nb_classes)
Y_test = np_utils.to_categorical(y_test, nb_classes)

model = Sequential()

model.add(Convolution2D(nb_filters, nb_conv, nb_conv,
                        border_mode='same',
                        input_shape=(1, img_rows, img_cols)))
model.add(Activation('relu'))
model.add(Convolution2D(nb_filters, nb_conv, nb_conv))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(nb_pool, nb_pool)))
model.add(Dropout(0.25))

model.add(Flatten())
model.add(Dense(128))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(nb_classes))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy', optimizer='adadelta')

layer_id = 0
model.extract_features = theano.function(
            [model.layers[0].input],
            model.layers[layer_id].get_output(False),
            on_unused_input='ignore', allow_input_downcast=True)

output = model.extract_features(X_test[0:1, :])
print("output size of layer : "+str(layer_id)+" has shape "+str(output.shape))

**Comment by jocicmarko** [comment]
You should check your layers' shapes by using **print_layer_shapes** function, found in keras/utils/layer_utils.

**Comment by EderSantana** [comment]
check nb_conv=3 if @jocicmarko suggestion doesn't work

**Comment by MinhazPalasara** [comment]
@EderSantana  it works as expected if the size of a filter is odd. Not for even.

So, I think the same border_mode implementation on gpu is right for odd kernel size.

**Comment by EderSantana** [comment]
@MinhazPalasara here is where we calculate the output size https://github.com/fchollet/keras/blob/master/keras/layers/convolutional.py#L16-L26

Do you mind giving suggestions for what your mean?

**Comment by jruales** [comment]
@EderSantana The problem seems to be here: https://github.com/fchollet/keras/blob/master/keras/layers/convolutional.py#L247 since symmetric padding is being applied to the input image, the same amount on each side of the image.

For a kernel that is even-sized along a dimension when using the "same" mode, it is necessary to have one more pixel of padding on one side of the image along this dimension in order to ensure that the output is of the correct size (currently, the image incorrectly shrinks in size along this dimension after the operation).

This problem seems to be only present in the GPU implementation, since the CPU implementation does ensure that the output has the correct size: https://github.com/fchollet/keras/blob/master/keras/layers/convolutional.py#L266
